### PR TITLE
Remote read first class: allow blocking queries over remote read

### PR DESCRIPTION
**Pull request by krajorama**
#### What this PR does

Implement blocking of queries via the per tenant configuration `blocked_queries`.
How it's applied: the blocking is applied individually to each query in the remote read request.
The query is turned into a selector, for example:  `{__name__="up"}`, including the curly braces.
The blocking is applied to each query. If one is blocked, the whole request is blocked.

#### Which issue(s) this PR fixes or relates to

Related to https://github.com/grafana/mimir-squad/issues/2148 (internal)

#### Checklist

- [X] Tests updated.
- [X] Documentation added.
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- N/A [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by krajorama** [comment]
Error message from local test:
```
$ cmd/mimirtool/mimirtool remote-read dump --address=http://localhost:8007 --selector="foobar"
INFO[0000] Created remote read client using endpoint 'http://localhost:8007/prometheus/api/v1/read'
INFO[0000] Querying time from=2024-06-14T10:38:35+02:00 to=2024-06-14T11:38:35+02:00 with selector=foobar 
mimirtool: error: remote server http://localhost:8007/prometheus/api/v1/read returned HTTP status 400 Bad Request: "status":"error","errorType":"bad_data","error":"remote read error (matchers_0: {__name__=\"foobar\"}): the request has 
been blocked by the cluster administrator (err-mimir-query-blocked)"}
```

File: pkg/frontend/querymiddleware/roundtrip_test.go [hunk 1]
```diff
@@ -516,6 +517,34 @@ func TestRemoteReadMiddleware(t *testing.T) {
 			makeRequest: generateTestRemoteReadRequest,
 			limits:      mockLimits{},
 		},
+		"block an exact query": {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/roundtrip_test.go:520]: I'm not fully convinced we should test it here. Why not adding test cases to `Test_queryBlocker_Do`?

Comment (krajorama) [reply pkg/frontend/querymiddleware/roundtrip_test.go:617]: For end-to-end test, I'll add a test in `integration`

File: docs/sources/mimir/configure/configure-blocked-queries.md [hunk 2]
```diff
@@ -24,6 +24,12 @@ overrides:
         regex: true
 ```
 
+The blocking is enforced on instant and range queries as well as remote read queries.
+
+For instant and range queries the pattern is evaluated against the query, for remote read requests, the pattern is evaluated against each set of matchers, as if the matchers formed a vector selector. If any set of matchers is blocked, the whole remote read request is rejected.
+
+For example the remote read query that contains the matcher `__name__` regex matched to `foo.*` is interpreted as `{__name__=~"foo.*"}`. To restrict the blocking to such selectors, include the curly braces in your pattern, e.g. `{.*foo.*}`.
```
Comment (pracucci) [inline docs/sources/mimir/configure/configure-blocked-queries.md:31]: > e.g. `{.*foo.*}`

A nit: if this is a regex, then the curly brackets should be escaped.

**Review by pracucci — APPROVED** [review]
Nice job, LGTM (modulo a nit).

File: docs/sources/mimir/configure/configure-blocked-queries.md (hunk 2 above)
Comment (krajorama) [reply docs/sources/mimir/configure/configure-blocked-queries.md]: nice catch

**Pull request merged**